xla/service/reduce_scatter_all_gather_combiner_test.cc

```diff
@@ -0,0 +1,161 @@
+        /*replica_count=*/num_replicas,
+        /*num_partitions=*/num_partitions);
+    config.set_use_spmd_partitioning(num_partitions > 1);
+    TF_ASSIGN_OR_RETURN(auto module,
+                        ParseAndReturnVerifiedModule(hlo_module, config));
+
+    auto changed = ReduceScatterAllGatherCombiner().Run(module.get());
+    if (!changed.ok()) {
+      return changed.status();
+    }
+    EXPECT_EQ(changed.value(), expect_change);
+    return StatusOr<std::unique_ptr<HloModule>>(std::move(module));
+  }
+
+  template <HloOpcode oc>
+  size_t CollectiveCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            HloPredicateIsOp<oc>);
+  }
+};
+
+TEST_F(ReduceScatterAllGatherCombinerTest, BranchesOptimized) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
+param.2 = bf16[8,128,1024]{2,1,0} parameter(1)
+reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+reduce-scatter.2 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.2), channel_id=9, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={1}, to_apply=add
+all-gather.2 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.2), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
+add.1 = bf16[8,128,1024]{2,1,0} add(all-gather.1, all-gather.2)
+}
+)";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, RunPass(hlo_string,
+                                               /*num_replicas=*/8,
+                                               /*num_partitions=*/1,
+                                               /*expect_change=*/true));
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllGather>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kReduceScatter>(module), 0);
+  EXPECT_EQ(CollectiveCount<HloOpcode::kAllReduce>(module), 2);
+}
+
+TEST_F(ReduceScatterAllGatherCombinerTest, IncompatibleProperties) {
+  absl::string_view hlo_string = R"(
+HloModule ReduceScatter
+
+add {
+  x = bf16[] parameter(0)
+  y = bf16[] parameter(1)
+  ROOT add = bf16[] add(x, y)
+}
+
+ENTRY main {
+param.1 = bf16[8,128,1024]{2,1,0} parameter(0)
+reduce-scatter.1 = bf16[8,64,1024]{2,1,0} reduce-scatter(param.1), channel_id=8, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=false, dimensions={1}, to_apply=add
+all-gather.1 = bf16[8,128,1024]{2,1,0} all-gather(reduce-scatter.1), channel_id=5, replica_groups={{0,1},{2,3},{4,5},{6,7}}, dimensions={1}, use_global_device_ids=true
```

Comment:
Wonder if we can add more pairs with other properties different (i.e., everything same except one) for completeness?

Reply:
The test commented above was for the same purpose. The property that I found not being checked within the verifier was the `use_global_device_ids` param which I added the `IncompatibleProperties` test, Any other one you'd want me to test?